src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/AsyncMutex.cs

```diff
@@ -0,0 +1,230 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Threading.Tasks;
+
+namespace System.Threading
+{
+    /// <summary>Provides an async mutex.</summary>
+    /// <remarks>
+    /// This could be achieved with a <see cref="SemaphoreSlim"/> constructed with an initial
+    /// and max limit of 1.  However, this implementation is optimized to the needs of ManagedWebSocket,
+    /// which is that we expect zero contention in typical use cases.
+    /// </remarks>
+    internal sealed class AsyncMutex
```

Comment:
Note this is just putting back the AsyncMutex that was removed (due to no longer being necessary for the use case) in https://github.com/dotnet/runtime/pull/39166/files#diff-bd308ac282a26fd004ac30d3c41eee3d90338f4b7a1c7bdfdb8b8488345490e5L107.
The only change from that is I simplified the Waiter type to just be a TaskCompletionSource, and removed all of the pooling that was being done, since we don't expect contention here that warrants that additional complexity.